File: epoll.c
#define _GNU_SOURCE
#include <sys/socket.h>
#include <sys/epoll.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <stddef.h>
#include <unistd.h>
#include <pthread.h>

#define MAX_EVENTS 2048

char Http200[] = "HTTP/1.1 200 OK\r\n"
                 "Content-Length: 2\r\n"
                 "\r\n"
                 "{}";

void listenAndServe() {
	int srvSock = socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
	if (srvSock == -1) {
		return;
	}
	int one = 1;
	if (setsockopt(srvSock, SOL_SOCKET, SO_REUSEPORT, &one, sizeof(one)) == -1) {
		return;
	}
	if (setsockopt(srvSock, SOL_TCP, TCP_NODELAY, &one, sizeof(one)) == -1) {
		return;
	}
	if (setsockopt(srvSock, SOL_TCP, TCP_DEFER_ACCEPT, &one, sizeof(one)) == -1) {
		return;
	}
	int queueLen = 2048;
	if (setsockopt(srvSock, SOL_TCP, TCP_FASTOPEN, &queueLen, sizeof(queueLen)) == -1) {
		return;
	}
	struct sockaddr_in srvAddr = { .sin_family = AF_INET, .sin_addr.s_addr = INADDR_ANY, .sin_port = htons(80)};
	if (bind(srvSock, (struct sockaddr *) &srvAddr, sizeof(srvAddr)) == -1) {
		return;
	}
	if (listen(srvSock, SOMAXCONN) == -1) {
		return;
	}
	int epollFd = epoll_create1(0);
	if (epollFd == -1) {
		return;
	}
	struct epoll_event event = { .events = EPOLLIN, .data.fd = srvSock};
	if (epoll_ctl(epollFd, EPOLL_CTL_ADD, srvSock, &event) == -1) {
		return;
	}
	struct epoll_event events[MAX_EVENTS];
	char buf[16384];
	while(1) {
		int n = epoll_wait(epollFd, events, MAX_EVENTS, 0);
		if (n == -1) {
			return;
		}
		while(n > 0) {
			n--;
			int sock = events[n].data.fd;
			if (sock == srvSock) {
				sock = accept4(srvSock, NULL, NULL, SOCK_NONBLOCK);
				if (sock == -1) {
					return;
				}
				if (setsockopt(sock, SOL_TCP, TCP_NODELAY, &one, sizeof(one)) == -1) {
					return;
				}
				event.data.fd = sock;
				if (epoll_ctl(epollFd, EPOLL_CTL_ADD, sock, &event) == -1) {
					return;
				}
			}
			int m = recv(sock, buf, sizeof(buf), 0);
			if (m == 0) {
				close(sock);
				continue;
			}
			if (m == -1) {
				close(sock);
				continue;
			}
			if (buf[0] == 'G' || buf[0] == 'P') {
				send(sock, Http200, sizeof(Http200) - 1, 0);
			}
		}
	}
}

void *serverWorker(void *args) {
        listenAndServe();
        return NULL;
}

int main() {
        pthread_t thr1, thr2, thr3;
        if (pthread_create(&thr1, NULL, serverWorker, NULL) != 0) {
                return -1;
        }
        if (pthread_create(&thr2, NULL, serverWorker, NULL) != 0) {
                return -1;
        }
        if (pthread_create(&thr3, NULL, serverWorker, NULL) != 0) {
                return -1;
        }
        pthread_join(thr1, NULL);
        pthread_join(thr2, NULL);
        pthread_join(thr3, NULL);
        return 0;
}

File: epoll.go
package main

import (
	"syscall"
)

var Http200 = []byte("HTTP/1.1 200 OK\r\n" +
	"Server: hlc\r\n" +
	"Content-Length: 2\r\n" +
	"\r\n" +
	"{}")

func listenAndServe() {
	srvSock, err := syscall.Socket(syscall.AF_INET, (syscall.SOCK_STREAM | syscall.SOCK_NONBLOCK), syscall.IPPROTO_TCP)
	if err != nil {
		panic(err)
	}
	if err = syscall.SetsockoptInt(srvSock, syscall.SOL_SOCKET, 0x0F /* SO_REUSEPORT */, 1); err != nil {
		panic(err)
	}
	if err = syscall.SetsockoptInt(srvSock, syscall.SOL_TCP, syscall.TCP_NODELAY, 1); err != nil {
		panic(err)
	}
	if err = syscall.SetsockoptInt(srvSock, syscall.SOL_TCP, syscall.TCP_DEFER_ACCEPT, 1); err != nil {
		panic(err)
	}
	if err = syscall.SetsockoptInt(srvSock, syscall.SOL_TCP, 0x17 /* TCP_FASTOPEN */, 2048 /* Queue length */); err != nil {
		panic(err)
	}
	if err = syscall.Bind(srvSock, &syscall.SockaddrInet4{Port: 80}); err != nil {
		panic(err)
	}
	if err = syscall.Listen(srvSock, syscall.SOMAXCONN); err != nil {
		panic(err)
	}
	epollFd, err := syscall.EpollCreate1(0)
	if err != nil {
		panic(err)
	}
	event := syscall.EpollEvent{Fd: int32(srvSock), Events: syscall.EPOLLIN}
	if err = syscall.EpollCtl(epollFd, syscall.EPOLL_CTL_ADD, srvSock, &event); err != nil {
		panic(err)
	}
	events := make([]syscall.EpollEvent, 2048)
	buf := make([]byte, 16384)
	for {
		n, err := syscall.EpollWait(epollFd, events, 0)
		if err != nil {
			panic(err)
		}
		for i := 0; i < n; i++ {
			sock := int(events[i].Fd)
			if sock == srvSock {
				if sock, _, err = syscall.Accept4(sock, syscall.SOCK_NONBLOCK); err != nil {
					panic(err)
				}
				if err = syscall.SetsockoptInt(sock, syscall.SOL_TCP, syscall.TCP_NODELAY, 1); err != nil {
					panic(err)
				}
				event = syscall.EpollEvent{Fd: int32(sock), Events: syscall.EPOLLIN}
				if err = syscall.EpollCtl(epollFd, syscall.EPOLL_CTL_ADD, sock, &event); err != nil {
					panic(err)
				}
			}
			m, err := syscall.Read(sock, buf)
			if m == 0 {
				syscall.Close(sock)
				continue
			}
			if err != nil {
				syscall.Close(sock)
				continue
			}
			if buf[0] == 'G' || buf[0] == 'P' {
				syscall.Write(sock, Http200)
			}
		}
	}
}

func main() {
	go listenAndServe()
	go listenAndServe()
	listenAndServe()
}
